zilliqa/src/contracts/deposit_v1.sol

```diff
@@ -0,0 +1,293 @@
+        address newImplementation
+    ) internal virtual override onlyOwner {}
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(
+        address initialOwner,
+        uint256 _minimumStake,
+        uint256 _maximumStakers,
+        uint64 _blocksPerEpoch,
+        InitialStaker[] memory initialStakers
+    ) public initializer {
+        __Deposit_init(initialOwner);
+        DepositStorage storage $ = _getDepositStorage();
+
+        $.minimumStake = _minimumStake;
+        $.maximumStakers = _maximumStakers;
+        $.blocksPerEpoch = _blocksPerEpoch;
+        $.latestComputedEpoch = currentEpoch();
+
+        for (uint i = 0; i < initialStakers.length; i++) {
+            InitialStaker memory initialStaker = initialStakers[i];
+            bytes memory blsPubKey = initialStaker.blsPubKey;
+            bytes memory peerId = initialStaker.peerId;
+            address rewardAddress = initialStaker.rewardAddress;
+            address controlAddress = initialStaker.controlAddress;
+            uint256 amount = initialStaker.amount;
+
+            require(blsPubKey.length == 48);
+            require(peerId.length == 38);
+            require(
+                controlAddress != address(0),
+                "control address cannot be zero"
+            );
+
+            Committee storage currentCommittee = committee();
+            require(
+                currentCommittee.stakerKeys.length < $.maximumStakers,
+                "too many stakers"
+            );
+
+            Staker storage staker = $._stakersMap[blsPubKey];
+            // This must be a new staker, meaning the control address must be zero.
+            require(
+                staker.controlAddress == address(0),
+                "staker already exists"
+            );
+
+            if (amount < $.minimumStake) {
+                revert("stake is less than minimum stake");
+            }
+
+            $._stakerKeys[controlAddress] = blsPubKey;
+            staker.peerId = peerId;
+            staker.rewardAddress = rewardAddress;
+            staker.controlAddress = controlAddress;
+
+            currentCommittee.totalStake += amount;
+            currentCommittee.stakers[blsPubKey].balance = amount;
+            currentCommittee.stakers[blsPubKey].index =
+                currentCommittee.stakerKeys.length +
+                1;
+            currentCommittee.stakerKeys.push(blsPubKey);
+        }
+    }
+
+    // automatically incrementing the version number allows for
+    // upgrading the contract without manually specifying the next
+    // version number in the source file - use with caution since
+    // it won't be possible to identify the actual version of the
+    // source file without a hardcoded version number, but storing
+    // the file versions in separate folders would help
+    function reinitialize() public reinitializer(version() + 1) {}
```

Comment:
The reinitializer is only needed in v2 and subsequent versions, we can remove the `reinitialize()` function in v1.

Reply:
We will have to include all versions of the contract into every release and upgrade to the next higher version at the specified block heights as you mentioned above. Therefore I'd suggest to use hardcoded version numbers i.e. `deposit_v<n>.sol` in the file name and `reinitializer(<n>)` in the modifier i.e. `reinitilizer(2)` above.